File: web/src/lib/review/device-proof.ts
import { ed25519 } from '@noble/curves/ed25519.js';
import { base64UrlDecode, base64UrlEncode, toCanonicalBytes } from './browser-crypto';
import type { MailboxEnvelope } from './browser-ws';

export const DEVICE_WS_PROOF_PURPOSE_V3 = 'attn device websocket proof v3';
export const DEVICE_SIGNAL_PROOF_PURPOSE_V3 = 'attn device signal proof v3';
export const DEVICE_HTTP_PROOF_PURPOSE_V3 = 'attn device http proof v3';
export const DEVICE_PROOF_LIFETIME_MS = 60_000;

export interface DeviceWebSocketProofV3 {
  expiresAt: number;
  nonce: string;
  signature: string;
}

export interface DeviceSignalProofInputV3 {
  roomId: string;
  envelopeId: string;
  authorId: string;
  deviceId: string;
  targetDeviceId: string | null;
  signalClass?: 'presence';
  generation: number;
  createdAt: number;
  expiresAt: number;
  nonce: string;
  ciphertext: string;
  ciphertextBytes: number;
}

export interface DeviceHttpProofInputV3 {
  resourceKind: 'room' | 'share';
  resourceId: string;
  deviceId: string;
  method: 'POST' | 'DELETE';
  path: string;
  bodySha256: string;
  bodyLength: number;
  powToken: string;
}

export function canonicalDeviceHttpProofV3(input: DeviceHttpProofInputV3): Uint8Array {
  return toCanonicalBytes({
    bodyLength: input.bodyLength,
    bodySha256: input.bodySha256,
    deviceId: input.deviceId,
    method: input.method,
    path: input.path,
    powToken: input.powToken,
    purpose: DEVICE_HTTP_PROOF_PURPOSE_V3,
    resourceId: input.resourceId,
    resourceKind: input.resourceKind,
    v: 3,
  });
}

export async function createDeviceHttpProofV3(input: {
  resourceKind: 'room' | 'share';
  resourceId: string;
  deviceId: string;
  method: 'POST' | 'DELETE';
  path: string;
  body: Uint8Array;
  powToken: string;
  signingSecret: Uint8Array;
}): Promise<string> {
  if (input.signingSecret.byteLength !== 32) throw new Error('device signing secret must be 32 bytes');
  if (input.resourceId.length === 0 || input.deviceId.length === 0 || input.path.length === 0 || input.powToken.length === 0) {
    throw new Error('device HTTP proof contains an empty bound field');
  }
  const digestBytes = new Uint8Array(await crypto.subtle.digest('SHA-256', ownedBuffer(input.body)));
  const canonical = canonicalDeviceHttpProofV3({
    resourceKind: input.resourceKind,
    resourceId: input.resourceId,
    deviceId: input.deviceId,
    method: input.method,
    path: input.path,
    bodySha256: base64UrlEncode(digestBytes),
    bodyLength: input.body.byteLength,
    powToken: input.powToken,
  });
  try {
    return base64UrlEncode(ed25519.sign(canonical, input.signingSecret));
  } finally {
    digestBytes.fill(0);
    canonical.fill(0);
  }
}

function ownedBuffer(bytes: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(bytes.byteLength);
  copy.set(bytes);
  return copy.buffer;
}

export function canonicalDeviceSignalProofV3(input: DeviceSignalProofInputV3): Uint8Array {
  return toCanonicalBytes({
    authorId: input.authorId,
    ciphertext: input.ciphertext,
    ciphertextBytes: input.ciphertextBytes,
    createdAt: input.createdAt,
    deviceId: input.deviceId,
    envelopeId: input.envelopeId,
    expiresAt: input.expiresAt,
    generation: input.generation,
    nonce: input.nonce,
    purpose: DEVICE_SIGNAL_PROOF_PURPOSE_V3,
    roomId: input.roomId,
    ...(input.signalClass === undefined ? {} : { signalClass: input.signalClass }),
    targetDeviceId: input.targetDeviceId,
    v: 3,
  });
}

export function signDeviceSignalProofV3(
  input: DeviceSignalProofInputV3,
  signingSecret: Uint8Array,
): string {
  if (signingSecret.length !== 32) throw new Error('device signing secret must be 32 bytes');
  validateDeviceSignalProofInput(input);
  const canonical = canonicalDeviceSignalProofV3(input);
  try {
    return base64UrlEncode(ed25519.sign(canonical, signingSecret));
  } finally {
    canonical.fill(0);
  }
}

export function verifyDeviceSignalProofV3(
  envelope: MailboxEnvelope,
  roomId: string,
  publicSigningKey: string,
): void {
  if (!Number.isSafeInteger(envelope.signalGeneration) || (envelope.signalGeneration ?? -1) < 0) {
    throw new Error('signal generation is missing or invalid');
  }
  if (typeof envelope.deviceSignature !== 'string') {
    throw new Error('signal device signature is missing');
  }
  const input = signalProofInputFromEnvelope(envelope, roomId);
  validateDeviceSignalProofInput(input);
  const signature = base64UrlDecode(envelope.deviceSignature);
  const publicKey = base64UrlDecode(publicSigningKey);
  if (signature.length !== 64 || publicKey.length !== 32) {
    throw new Error('signal device proof key or signature has invalid length');
  }
  const canonical = canonicalDeviceSignalProofV3(input);
  try {
    if (!ed25519.verify(signature, canonical, publicKey)) {
      throw new Error('signal device signature is invalid');
    }
  } finally {
    canonical.fill(0);
    signature.fill(0);
    publicKey.fill(0);
  }
}

export function signalProofInputFromEnvelope(
  envelope: MailboxEnvelope,
  roomId: string,
): DeviceSignalProofInputV3 {
  if (envelope.signalGeneration === undefined) throw new Error('signal generation is missing');
  return {
    roomId,
    envelopeId: envelope.envelopeId,
    authorId: envelope.authorId,
    deviceId: envelope.deviceId,
    targetDeviceId: envelope.target?.deviceId ?? null,
    ...(envelope.signalClass === undefined ? {} : { signalClass: envelope.signalClass }),
    generation: envelope.signalGeneration,
    createdAt: envelope.createdAt,
    expiresAt: envelope.expiresAt,
    nonce: envelope.nonce,
    ciphertext: envelope.ciphertext,
    ciphertextBytes: envelope.ciphertextBytes,
  };
}

function validateDeviceSignalProofInput(input: DeviceSignalProofInputV3): void {
  for (const value of [input.roomId, input.envelopeId, input.authorId, input.deviceId, input.nonce, input.ciphertext]) {
    if (value.length === 0) throw new Error('signal proof contains an empty bound field');
  }
  if (!Number.isSafeInteger(input.generation) || input.generation < 0) throw new Error('signal generation is invalid');
  if (!Number.isSafeInteger(input.createdAt) || input.createdAt <= 0) throw new Error('signal createdAt is invalid');
  if (!Number.isSafeInteger(input.expiresAt) || input.expiresAt < input.createdAt) throw new Error('signal expiresAt is invalid');
  if (!Number.isSafeInteger(input.ciphertextBytes) || input.ciphertextBytes <= 0) throw new Error('signal ciphertext length is invalid');
}

export function canonicalDeviceWebSocketProofV3(input: {
  roomId: string;
  deviceId: string;
  path: string;
  expiresAt: number;
  nonce: string;
}): Uint8Array {
  return toCanonicalBytes({
    deviceId: input.deviceId,
    expiresAt: input.expiresAt,
    method: 'GET',
    nonce: input.nonce,
    path: input.path,
    purpose: DEVICE_WS_PROOF_PURPOSE_V3,
    roomId: input.roomId,
    v: 3,
  });
}

export function createDeviceWebSocketProofV3(input: {
  roomId: string;
  deviceId: string;
  path: string;
  signingSecret: Uint8Array;
  now?: number;
  nonceBytes?: Uint8Array;
}): DeviceWebSocketProofV3 {
  if (input.signingSecret.length !== 32) throw new Error('device signing secret must be 32 bytes');
  const nonceBytes = input.nonceBytes === undefined
    ? crypto.getRandomValues(new Uint8Array(16))
    : new Uint8Array(input.nonceBytes);
  if (nonceBytes.length !== 16) throw new Error('device proof nonce must be 16 bytes');
  const expiresAt = (input.now ?? Date.now()) + DEVICE_PROOF_LIFETIME_MS;
  const nonce = base64UrlEncode(nonceBytes);
  const canonical = canonicalDeviceWebSocketProofV3({
    roomId: input.roomId,
    deviceId: input.deviceId,
    path: input.path,
    expiresAt,
    nonce,
  });
  try {
    return {
      expiresAt,
      nonce,
      signature: base64UrlEncode(ed25519.sign(canonical, input.signingSecret)),
    };
  } finally {
    canonical.fill(0);
    nonceBytes.fill(0);
  }
}

File: src/review/crypto/device_proof.rs
use base64::Engine as _;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use serde::Serialize;

use super::canonical::{CanonError, to_canonical_bytes};
use super::signing::{DeviceSigningKey, DeviceVerifyingKey, SignError};

pub const DEVICE_WS_PROOF_PURPOSE_V3: &str = "attn device websocket proof v3";
pub const DEVICE_SIGNAL_PROOF_PURPOSE_V3: &str = "attn device signal proof v3";
pub const DEVICE_PROOF_LIFETIME_MS: u64 = 60_000;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeviceWebSocketProofV3 {
    pub expires_at: u64,
    pub nonce: String,
    pub signature: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct CanonicalDeviceWebSocketProofV3<'a> {
    device_id: &'a str,
    expires_at: u64,
    method: &'static str,
    nonce: &'a str,
    path: &'a str,
    purpose: &'static str,
    room_id: &'a str,
    v: u32,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct CanonicalDeviceSignalProofV3<'a> {
    author_id: &'a str,
    ciphertext: &'a str,
    ciphertext_bytes: u64,
    created_at: u64,
    device_id: &'a str,
    envelope_id: &'a str,
    expires_at: u64,
    generation: u64,
    nonce: &'a str,
    purpose: &'static str,
    room_id: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    signal_class: Option<&'a str>,
    target_device_id: Option<&'a str>,
    v: u32,
}

#[allow(clippy::too_many_arguments)]
pub fn canonical_device_signal_proof_v3(
    room_id: &str,
    envelope_id: &str,
    author_id: &str,
    device_id: &str,
    target_device_id: Option<&str>,
    signal_class: Option<&str>,
    generation: u64,
    created_at: u64,
    expires_at: u64,
    nonce: &str,
    ciphertext: &str,
    ciphertext_bytes: u64,
) -> Result<Vec<u8>, CanonError> {
    to_canonical_bytes(&CanonicalDeviceSignalProofV3 {
        author_id,
        ciphertext,
        ciphertext_bytes,
        created_at,
        device_id,
        envelope_id,
        expires_at,
        generation,
        nonce,
        purpose: DEVICE_SIGNAL_PROOF_PURPOSE_V3,
        room_id,
        signal_class,
        target_device_id,
        v: 3,
    })
}

#[allow(clippy::too_many_arguments)]
pub fn sign_device_signal_proof_v3(
    signing_key: &DeviceSigningKey,
    room_id: &str,
    envelope_id: &str,
    author_id: &str,
    device_id: &str,
    target_device_id: Option<&str>,
    signal_class: Option<&str>,
    generation: u64,
    created_at: u64,
    expires_at: u64,
    nonce: &str,
    ciphertext: &str,
    ciphertext_bytes: u64,
) -> Result<String, CanonError> {
    let canonical = canonical_device_signal_proof_v3(
        room_id,
        envelope_id,
        author_id,
        device_id,
        target_device_id,
        signal_class,
        generation,
        created_at,
        expires_at,
        nonce,
        ciphertext,
        ciphertext_bytes,
    )?;
    Ok(URL_SAFE_NO_PAD.encode(signing_key.sign_protocol_bytes(&canonical)))
}

#[allow(clippy::too_many_arguments)]
pub fn verify_device_signal_proof_v3(
    verifying_key: &DeviceVerifyingKey,
    signature: &str,
    room_id: &str,
    envelope_id: &str,
    author_id: &str,
    device_id: &str,
    target_device_id: Option<&str>,
    signal_class: Option<&str>,
    generation: u64,
    created_at: u64,
    expires_at: u64,
    nonce: &str,
    ciphertext: &str,
    ciphertext_bytes: u64,
) -> Result<(), String> {
    let signature_bytes = URL_SAFE_NO_PAD
        .decode(signature)
        .map_err(|error| format!("signal signature base64url: {error}"))?;
    let signature: [u8; 64] = signature_bytes.try_into().map_err(|bytes: Vec<u8>| {
        format!("signal signature must be 64 bytes, got {}", bytes.len())
    })?;
    let canonical = canonical_device_signal_proof_v3(
        room_id,
        envelope_id,
        author_id,
        device_id,
        target_device_id,
        signal_class,
        generation,
        created_at,
        expires_at,
        nonce,
        ciphertext,
        ciphertext_bytes,
    )
    .map_err(|error| format!("signal proof canonicalization: {error}"))?;
    verifying_key
        .verify_protocol_bytes(&canonical, &signature)
        .map_err(|error: SignError| format!("signal device signature: {error}"))
}

pub fn canonical_device_websocket_proof_v3(
    room_id: &str,
    device_id: &str,
    path: &str,
    expires_at: u64,
    nonce: &str,
) -> Result<Vec<u8>, CanonError> {
    to_canonical_bytes(&CanonicalDeviceWebSocketProofV3 {
        device_id,
        expires_at,
        method: "GET",
        nonce,
        path,
        purpose: DEVICE_WS_PROOF_PURPOSE_V3,
        room_id,
        v: 3,
    })
}

pub fn create_device_websocket_proof_v3(
    signing_key: &DeviceSigningKey,
    room_id: &str,
    device_id: &str,
    path: &str,
    now_ms: u64,
    nonce_bytes: &[u8; 16],
) -> Result<DeviceWebSocketProofV3, CanonError> {
    let expires_at = now_ms.saturating_add(DEVICE_PROOF_LIFETIME_MS);
    let nonce = URL_SAFE_NO_PAD.encode(nonce_bytes);
    let canonical =
        canonical_device_websocket_proof_v3(room_id, device_id, path, expires_at, &nonce)?;
    Ok(DeviceWebSocketProofV3 {
        expires_at,
        nonce,
        signature: URL_SAFE_NO_PAD.encode(signing_key.sign_protocol_bytes(&canonical)),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_shape_and_signature_are_pinned() {
        let key = DeviceSigningKey::from_bytes(&[0x11; 32]).unwrap();
        let proof = create_device_websocket_proof_v3(
            &key,
            "room-vector",
            "device-vector",
            "/v3/rooms/room-vector/socket",
            1_700_000_000_000,
            &[0x22; 16],
        )
        .unwrap();
        assert_eq!(proof.expires_at, 1_700_000_060_000);
        assert_eq!(proof.nonce, "IiIiIiIiIiIiIiIiIiIiIg");
        assert_eq!(
            proof.signature,
            "WoSwnColLautRZzjGUU2M9h0Fj2Tjz1uS2d2kqEISDfl-xLs8YpjBBQZG5ddK4EsRdCblGiio6QlT8qjMDMcCQ"
        );
        let canonical = canonical_device_websocket_proof_v3(
            "room-vector",
            "device-vector",
            "/v3/rooms/room-vector/socket",
            proof.expires_at,
            &proof.nonce,
        )
        .unwrap();
        assert_eq!(
            String::from_utf8(canonical).unwrap(),
            "{\"deviceId\":\"device-vector\",\"expiresAt\":1700000060000,\"method\":\"GET\",\"nonce\":\"IiIiIiIiIiIiIiIiIiIiIg\",\"path\":\"/v3/rooms/room-vector/socket\",\"purpose\":\"attn device websocket proof v3\",\"roomId\":\"room-vector\",\"v\":3}"
        );
    }

    #[test]
    fn signal_vector_matches_browser_and_binds_target_and_key() {
        let key = DeviceSigningKey::from_bytes(&[0x11; 32]).unwrap();
        let signature = sign_device_signal_proof_v3(
            &key,
            "room-vector",
            "envelope-vector",
            "author-vector",
            "device-vector",
            Some("target-vector"),
            None,
            7,
            1_700_000_000_000,
            1_700_003_600_000,
            "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIi",
            "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM",
            32,
        )
        .unwrap();
        assert_eq!(
            signature,
            "PbzT2GYKbUkXTMr8VdpNa-cGkfLXk8vZPOLF4C3fDZJij83iE7Aea4lQbhA1BFJlZGg-tRI2Fr_IgbNK4jzXAQ"
        );
        let verify = |verifying_key: &DeviceVerifyingKey, target: Option<&str>| {
            verify_device_signal_proof_v3(
                verifying_key,
                &signature,
                "room-vector",
                "envelope-vector",
                "author-vector",
                "device-vector",
                target,
                None,
                7,
                1_700_000_000_000,
                1_700_003_600_000,
                "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIi",
                "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM",
                32,
            )
        };
        verify(&key.verifying_key(), Some("target-vector")).unwrap();
        assert!(verify(&key.verifying_key(), Some("rewritten")).is_err());
        let wrong = DeviceSigningKey::from_bytes(&[0x44; 32]).unwrap();
        assert!(verify(&wrong.verifying_key(), Some("target-vector")).is_err());
    }

    #[test]
    fn signal_proof_binds_replaceable_presence_class() {
        let key = DeviceSigningKey::from_bytes(&[0x21; 32]).unwrap();
        let signature = sign_device_signal_proof_v3(
            &key,
            "room-presence",
            "envelope-presence",
            "author-presence",
            "device-presence",
            None,
            Some("presence"),
            9,
            1_700_000_000_009,
            1_700_003_600_009,
            "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIi",
            "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM",
            32,
        )
        .unwrap();
        assert_eq!(
            signature,
            "abvJIj3s-e19-N7f0ZH9B7skWw1YbrDGePpkeiOxr-aTSts2jRoiltjAmigTlV57HlLL6QGsWCGnIjuJh3TpDA"
        );
        verify_device_signal_proof_v3(
            &key.verifying_key(),
            &signature,
            "room-presence",
            "envelope-presence",
            "author-presence",
            "device-presence",
            None,
            Some("presence"),
            9,
            1_700_000_000_009,
            1_700_003_600_009,
            "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIi",
            "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM",
            32,
        )
        .unwrap();
        assert!(
            verify_device_signal_proof_v3(
                &key.verifying_key(),
                &signature,
                "room-presence",
                "envelope-presence",
                "author-presence",
                "device-presence",
                None,
                None,
                9,
                1_700_000_000_009,
                1_700_003_600_009,
                "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIi",
                "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM",
                32,
            )
            .is_err()
        );
    }
}
